contracts/stake/src/wasm.rs

```diff
@@ -0,0 +1,182 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::*;
+
+use alloc::vec::Vec;
+use canonical::{Canon, Sink, Source};
+use dusk_abi::{ContractState, ReturnValue, Transaction};
+use dusk_bls12_381::BlsScalar;
+use dusk_bytes::Serializable;
+use dusk_pki::PublicKey;
+use dusk_schnorr::Signature;
+use phoenix_core::Note;
+use transfer_contract::Call;
+
+pub const TX_STAKE: u8 = 0x00;
+pub const TX_EXTEND: u8 = 0x01;
+pub const TX_WITHDRAW: u8 = 0x02;
+pub const TX_SLASH: u8 = 0x03;
+
+const PAGE_SIZE: usize = 1024 * 32;
+
+#[no_mangle]
+fn q(bytes: &mut [u8; PAGE_SIZE]) {
+    let mut source = Source::new(&bytes[..]);
+
+    let _contract =
+        StakeContract::decode(&mut source).expect("Failed to read state");
+    let qid = u8::decode(&mut source).expect("Failed to read query ID");
+
+    let mut sink = Sink::new(&mut bytes[..]);
+
+    let ret = match qid {
+        rusk_abi::PAYMENT_INFO => ReturnValue::from_canon(&PAYMENT_INFO),
+
+        _ => panic!("Undefined query ID"),
+    };
+
+    ret.encode(&mut sink);
+}
+
+#[no_mangle]
+fn t(bytes: &mut [u8; PAGE_SIZE]) {
+    let mut source = Source::new(bytes);
+
+    let mut contract =
+        StakeContract::decode(&mut source).expect("Failed to read state");
+    let tid = u8::decode(&mut source).expect("Failed to read tx ID");
+
+    match tid {
+        TX_STAKE => {
+            let (pk, value, spend_proof): (PublicKey, u64, Vec<u8>) =
+                Canon::decode(&mut source).expect("Failed to parse arguments");
+
+            contract.stake(pk, value, spend_proof);
+        }
+
+        TX_EXTEND => {
+            let (pk, signature): (PublicKey, Signature) =
+                Canon::decode(&mut source).expect("Failed to parse arguments");
+
+            contract.extend(pk, signature);
+        }
+
+        TX_WITHDRAW => {
+            let (pk, signature, note, withdraw_proof): (
+                PublicKey,
+                Signature,
+                Note,
+                Vec<u8>,
+            ) = Canon::decode(&mut source).expect("Failed to parse arguments");
+
+            contract.withdraw(pk, signature, note, withdraw_proof);
+        }
+
+        TX_SLASH => panic!("The arbitrationContract is yet to be defined."),
```

Comment:
The slash should be still implemented at least for the branch where the `arbitrationContract` is not involved (see: https://app.gitbook.com/@dusk-network/s/specs/specifications/smart-contracts/genenesis-contracts/stake-contract-1/).

Edit: I spoke with @toghrulmaharramov and the idea would be ignore the `if` at all currently and burn always the token (the `else` in the documentation).

Reply:
Done